File: bilmarknad-mcp/src/bilmarknad_mcp/kvd.py
from __future__ import annotations

from typing import Any

import httpx

from bilmarknad_mcp.schema import CarListing

KVD_BASE = "https://www.kvd.se"
CANDIDATE_PATHS = (
    "/api/vehicles",
    "/api/search",
    "/api/v1/vehicles",
)


class KvdUnavailableError(RuntimeError):
    pass


class KvdClient:
    def __init__(self, client: httpx.Client | None = None):
        self._client = client
        self._owns = client is None
        self._available: bool | None = None

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(timeout=20.0, headers={"User-Agent": "bilmarknad-mcp/0.1"})
        return self._client

    def probe(self) -> bool:
        if self._available is not None:
            return self._available
        client = self._get_client()
        for path in CANDIDATE_PATHS:
            try:
                response = client.get(KVD_BASE + path, params={"limit": 1})
            except httpx.HTTPError:
                continue
            if response.status_code == 200 and "application/json" in response.headers.get(
                "content-type", ""
            ):
                self._available = True
                return True
        self._available = False
        return False

    def search(self, **kwargs: Any) -> list[CarListing]:
        if not self.probe():
            raise KvdUnavailableError("No public KVD API endpoint responded with JSON")
        return []

    def close(self) -> None:
        if self._owns and self._client is not None:
            self._client.close()

File: internal/bilmarknad/kvd/kvd.go
package kvd

import (
	"context"
	"fmt"
	"net/http"
	"strings"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
)

const baseURL = "https://www.kvd.se"

var candidatePaths = []string{
	"/api/vehicles",
	"/api/search",
	"/api/v1/vehicles",
}

// UnavailableError indicates no public KVD API endpoint responded.
type UnavailableError struct{}

func (UnavailableError) Error() string {
	return "no public KVD API endpoint responded with JSON"
}

// Client probes KVD for a public JSON API.
type Client struct {
	httpClient *http.Client
	owns       bool
	available  *bool
}

func NewClient(c *http.Client) *Client {
	if c == nil {
		c = httputil.NewClient()
	}
	return &Client{httpClient: c, owns: true}
}

func (c *Client) Probe(ctx context.Context) bool {
	if c.available != nil {
		return *c.available
	}
	for _, path := range candidatePaths {
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, baseURL+path, nil)
		if err != nil {
			continue
		}
		q := req.URL.Query()
		q.Set("limit", "1")
		req.URL.RawQuery = q.Encode()
		resp, err := c.httpClient.Do(req)
		if err != nil {
			continue
		}
		resp.Body.Close()
		if resp.StatusCode == 200 && strings.Contains(resp.Header.Get("Content-Type"), "application/json") {
			v := true
			c.available = &v
			return true
		}
	}
	v := false
	c.available = &v
	return false
}

func (c *Client) Search(ctx context.Context) ([]schema.CarListing, error) {
	if !c.Probe(ctx) {
		return nil, UnavailableError{}
	}
	return nil, nil
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}

func (c *Client) String() string { return fmt.Sprintf("kvd.Client") }
